src/main/java/com/pusher/client/channel/impl/PresenceChannelImpl.java

```diff
@@ -65,35 +64,10 @@ else if (event.equals(MEMBER_REMOVED_EVENT)) {
     }
 
     @Override
-    @SuppressWarnings("rawtypes")
     public String toSubscribeMessage() {
-
-        final String authResponse = getAuthResponse();
-
-        try {
-            final Map authResponseMap = GSON.fromJson(authResponse, Map.class);
-            final String authKey = (String)authResponseMap.get("auth");
-            final Object channelData = authResponseMap.get("channel_data");
-
-            storeMyUserId(channelData);
-
-            final Map<Object, Object> jsonObject = new LinkedHashMap<Object, Object>();
-            jsonObject.put("event", "pusher:subscribe");
-
-            final Map<Object, Object> dataMap = new LinkedHashMap<Object, Object>();
-            dataMap.put("channel", name);
-            dataMap.put("auth", authKey);
-            dataMap.put("channel_data", channelData);
-
-            jsonObject.put("data", dataMap);
-
-            final String json = GSON.toJson(jsonObject);
-
-            return json;
-        }
-        catch (final Exception e) {
-            throw new AuthorizationFailureException("Unable to parse response from Authorizer: " + authResponse, e);
-        }
+        String msg = super.toSubscribeMessage();
+        storeMyUserId(channelData);
```

Comment:
Previously, if the channel data was missing (e.g. if someone tried to authorise a presence channel on an auth endpoint only configured to authorise private channels) then it looks like there would have been a null pointer exception thrown in `storeMyUserId` which would have been caught and converted to an `AuthorizationFailureException` above. It looks like now there would just be a null pointer exception. I wonder if it might be better to wrap the call to `storeMyUserId`  in a try catch like before.

Reply:
Great catch!

Reply:
Have a look at the latest commit. It should now throw `AuthorizationFailureException`s like before. I also moved the side effect of storing the user id into `toSubscribeMessage`, which I think is cleaner than storing it in a method that is responsible for parsing. I renamed `storeMyUserId` to `extractUserIdFromChannelData` because of this.

`extractUserIdFromChannelData` throws `AuthorizationFailureException` which I think is a bit unclean since it's a fairly generic method that could in theory could have other users. It might be nicer to have a `ParseException` which gets wrapped into a `AuthorizationFailureException` higher up. I don't think this is worth doing now though.